src/deterministicArchive.js

```diff
@@ -0,0 +1,129 @@
+import { Writable } from 'stream';
+import crypto from 'crypto';
+import fs from 'fs';
+import path from 'path';
+
+import Archiver from 'archiver';
+
+import validateArchive from './validateArchive';
+
+// We're setting the creation date to the same for all files so that the zip
+// packages created for the same content ends up having the same fingerprint.
+const FILE_CREATION_DATE = new Date('Fri Feb 08 2019 13:31:55 GMT+0100 (CET)');
+
+/**
+ * Resolves all files in a directory and all of its subdirectories
+ *
+ * @param {string} dirOrFile
+ * @returns {Promise<Array<{name: string, path: string}>>}
+ */
+async function resolveFilesRecursiveForDir(dirOrFile) {
+  const isDir = (await fs.promises.lstat(dirOrFile)).isDirectory();
+
+  if (isDir) {
+    const files = await fs.promises.readdir(dirOrFile, {
+      withFileTypes: true,
+      recursive: true,
```

Comment:
Node 18 adds support for recursive: true here, so we can use that and simplify our directory scanning.